File: Sources/XMTP/Messages/Envelope.swift
//
//  Envelope.swift
//
//
//  Created by Pat Nakajima on 11/26/22.
//

import Foundation
import XMTPProto

public typealias Envelope = Xmtp_MessageApi_V1_Envelope

extension Envelope {
	init(topic: Topic, timestamp: Date, message: Data) {
		self.init()
		contentTopic = topic.description
		timestampNs = UInt64(timestamp.millisecondsSinceEpoch * 1_000_000)
		self.message = message
	}

	init(topic: String, timestamp: Date, message: Data) {
		self.init()
		contentTopic = topic
		timestampNs = UInt64(timestamp.millisecondsSinceEpoch * 1_000_000)
		self.message = message
	}
}

File: library/src/main/java/org/xmtp/android/library/messages/Envelope.kt
package org.xmtp.android.library.messages

import com.google.protobuf.kotlin.toByteString
import java.util.Date

typealias Envelope = org.xmtp.proto.message.api.v1.MessageApiOuterClass.Envelope

class EnvelopeBuilder {
    companion object {
        fun buildFromString(topic: String, timestamp: Date, message: ByteArray): Envelope {
            return Envelope.newBuilder().apply {
                contentTopic = topic
                timestampNs = (timestamp.time * 1_000_000)
                this.message = message.toByteString()
            }.build()
        }

        fun buildFromTopic(topic: Topic, timestamp: Date, message: ByteArray): Envelope {
            return Envelope.newBuilder().apply {
                contentTopic = topic.description
                timestampNs = (timestamp.time * 1_000_000)
                this.message = message.toByteString()
            }.build()
        }
    }
}
